hoot-core/src/main/cpp/hoot/core/io/OsmApiChangeset.cpp

```diff
@@ -1023,9 +1135,94 @@ bool XmlChangeset::fixElement(ChangesetTypeMap& map, long id, long version)
         //  Change was made
         success = true;
       }
+      //  Update the tags if they are missing
+      for (int i = 0; i < element->getTagCount(); ++i)
+      {
+        QString key = element->getTagKey(i);
+        if (tags.contains(key))
+          tags.remove(key);
+      }
+      //  Add in any tags that are missing
+      QXmlStreamAttributes attributes;
+      for (QMap<QString, QString>::iterator it = tags.begin(); it != tags.end(); ++it)
+        attributes.append("", it.key(), it.value());
+      if (attributes.size() > 0)
+        element->addTag(XmlObject("tag", QXmlStreamAttributes()));
     }
   }
   return success;
 }
 
+long XmlChangeset::getNextNodeId()
+{
+  //  Get a new node ID from the generator
+  long id = _idGen.createNodeId();
+  //  Make sure that ID isn't already being used in the changeset
+  while (_allNodes.find(id) != _allNodes.end())
+    id = _idGen.createNodeId();
+  return id;
+}
```

Comment:
Get a new node ID from the negative generator and check if it is already being used in the changeset.  If it is, go to the next and repeat.  This gives unique negative IDs.